src/groth16/aggregate/accumulator.rs

```diff
@@ -2,13 +2,126 @@
+    /// Circuit breaker to allow canceling all checks and marking the whole check as failed.
+    valid: Arc<AtomicBool>,
+    merge_send: Sender<PairingCheck<E>>,
+    valid_recv: Receiver<bool>,
+    /// Random number generator used for generating the random coefficients.
+    rng: Mutex<R>,
+    /// Ensures that the non randomized check is only added exactly once.
+    non_random_check_done: AtomicBool,
+}
+
+impl<E: Engine, R: rand::RngCore + Send> PairingChecks<E, R> {
+    pub fn new(rng: R) -> Self {
+        let (merge_send, merge_recv): (Sender<PairingCheck<E>>, Receiver<PairingCheck<E>>) =
+            bounded(10);
+        let (valid_send, valid_recv) = bounded(1);
+
+        let valid = Arc::new(AtomicBool::new(true));
+        let valid_copy = valid.clone();
+
+        rayon::spawn(move || {
+            let mut acc = PairingCheck::new();
+            while let Ok(tuple) = merge_recv.recv() {
+                // only do work as long as we know we are still valid
+                if valid_copy.load(SeqCst) {
+                    acc.merge(&tuple);
+                }
+            }
+            if valid_copy.load(SeqCst) {
+                valid_send.send(acc.verify()).expect("failed to send");
+            }
+        });
+
+        PairingChecks {
+            valid,
+            merge_send,
+            valid_recv,
+            rng: Mutex::new(rng),
+            non_random_check_done: AtomicBool::new(false),
+        }
+    }
+
+    /// Fails the whole check.
+    pub fn invalidate(&self) {
+        self.valid.store(false, SeqCst);
+    }
+
+    fn merge_pair(&self, result: E::Fqk, exp: E::Fqk, must_randomize: bool) {
+        self.merge(PairingCheck::from_pair(result, exp), must_randomize);
+    }
+
+    fn merge_miller_one(&self, result: E::Fqk, must_randomize: bool) {
+        self.merge(PairingCheck::from_miller_one(result), must_randomize);
+    }
+
+    pub fn merge_pairing_equation(&self, left: Vec<E::Fqk>, pair: (E::Fqk, E::Fqk)) {
+        let must_randomize = self.non_random_check_done.load(SeqCst);
+
+        for l in left.iter() {
+            self.merge_miller_one(*l, must_randomize);
+        }
+        self.merge_pair(pair.0, pair.1, must_randomize);
+    }
+
+    pub fn merge_miller_inputs<'a>(
+        &self,
+        it: &[(&'a E::G1Affine, &'a E::G2Affine)],
+        out: &'a E::Fqk,
+    ) {
+        let must_randomize = self.non_random_check_done.load(SeqCst);
+        let coeff = {
+            let rng: &mut R = &mut self.rng.lock().unwrap();
+            derive_non_zero::<E, _>(rng)
+        };
+        self.merge(
+            PairingCheck::from_miller_inputs(coeff, it, out),
+            must_randomize,
+        );
+    }
+
+    fn merge(&self, check: PairingCheck<E>, must_randomize: bool) {
+        if !check.randomized {
+            assert!(
+                !must_randomize,
+                "Cannot merge non-randomized check with must_randomize true."
```

Comment:
We can actually - it's just so much more expensive:
We can do `e(A,B)^r = T^r` when the inputs are `e(A,B)` `T` but these are all in G_T which makes the exponentiation very expensive.

Reply:
Okay, maybe it should say "refusing to" since we prefer to fail at development time when opting into these optimizations.